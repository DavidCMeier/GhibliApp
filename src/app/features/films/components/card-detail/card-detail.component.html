<div class="flex-content" *ngIf="film">
  <mat-card class="card-detail">
    <mat-card-title>{{film.title}}</mat-card-title>
    <mat-card-subtitle> {{film.director}} - {{film.release_date}}</mat-card-subtitle>
    <img mat-card-image [src]="film.movie_banner" [alt]="film.title + ' banner'">
    <mat-card-content>
      <p>{{'films.detail.original_title' | translate: {originalTitle: film.original_title} }}</p>
      <p>{{ 'films.detail.original_title_romanised' | translate:{originalTitleRomanised: film.original_title_romanised} }}</p>
      <p>{{film.description}}</p>
      <p>{{ 'films.detail.directed' | translate:{director: film.director} }}</p>
      <p>{{ 'films.detail.produced' | translate:{producer: film.producer} }}</p>
      <p>{{ 'films.detail.duration' | translate:{duration: film.running_time} }}</p>
      <p>{{ 'films.detail.rt_score' | translate:{score: film.rt_score} }}</p>
    </mat-card-content>
  </mat-card>
</div>
