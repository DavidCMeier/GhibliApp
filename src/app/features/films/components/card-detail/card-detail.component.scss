.card-detail{
  max-width: 28rem;
  box-shadow: 7px 5px 6px 0px rgb(0 0 0 / 40%);
}


