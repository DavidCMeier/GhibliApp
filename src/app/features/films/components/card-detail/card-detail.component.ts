import { Component, Input, OnInit } from '@angular/core';
import { Film } from "../../models/film.model";

@Component({
  selector: 'app-card-detail',
  templateUrl: './card-detail.component.html',
  styleUrls: ['./card-detail.component.scss']
})
export class CardDetailComponent implements OnInit {

  @Input() film!: Film;
  constructor() { }

  ngOnInit(): void {
  }

}
