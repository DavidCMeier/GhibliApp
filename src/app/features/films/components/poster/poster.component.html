<img class="poster" [src]="poster" [alt]="'films.poster.alt' | translate: {title: filmTitle}">
