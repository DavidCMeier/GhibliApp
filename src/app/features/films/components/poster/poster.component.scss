.poster{
  width: 100%;
  height: auto;
  border-radius: 4px;
  box-shadow: 7px 5px 6px 0 rgb(0 0 0 / 40%);
  max-width: 29rem;
}
