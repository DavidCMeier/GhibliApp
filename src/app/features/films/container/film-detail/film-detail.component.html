<ng-container *ngIf="film$ | async as film">
  <div class="flex-content">
    <app-card-detail [film]="film"></app-card-detail>
    <app-poster [poster]="film.image" [filmTitle]="film.title"></app-poster>
  </div>
</ng-container>
<app-loading *ngIf="loading$ |async"></app-loading>
