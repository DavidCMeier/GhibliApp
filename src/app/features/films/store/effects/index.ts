import { FilmsPageEffects } from "./films-page.effects";
import { FilmDetailEffects } from "./film-detail.effects";

export const effects = [FilmsPageEffects, FilmDetailEffects]

export * from './film-detail.effects'
export * from './films-page.effects'
