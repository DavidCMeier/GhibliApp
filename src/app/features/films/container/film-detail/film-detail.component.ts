import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from '@angular/core';
import { FilmService } from "../../services/film.service";
import { Observable } from "rxjs";
import { Film } from "../../models/film.model";
import { ActivatedRoute } from "@angular/router";
import { Store } from "@ngrx/store";
import * as fromStore from "../../store";
import * as fromCoreStore from "../../../../core/store";
import { take, withLatestFrom } from "rxjs/operators";

@Component({
  selector: 'app-film-detail',
  templateUrl: './film-detail.component.html',
  styleUrls: ['./film-detail.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FilmDetailComponent implements OnInit, OnDestroy {

  film$ = this.store$.select(fromStore.getFilmDetail);
  loading$: Observable<boolean> = this.store$.select(fromStore.getFilmLoading);
  constructor(private store$: Store<fromStore.FilmsState>) { }

  ngOnInit(): void {
    this.store$
      .select(fromCoreStore.getRouteParams)
      .pipe(take(1), withLatestFrom(this.store$.select(fromCoreStore.getQueryParams)))
      .subscribe(([params, queryParams]) => {
        const uuid = params.id;
        console.log(queryParams)
        this.store$.dispatch(fromStore.loadFilm({uuid}));
      })
  }

  ngOnDestroy(): void {
    this.store$.dispatch(fromStore.unloadFilm());
  }

}
