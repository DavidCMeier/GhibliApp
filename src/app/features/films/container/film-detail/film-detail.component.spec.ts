import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FilmDetailComponent } from './film-detail.component';
import { RouterTestingModule } from "@angular/router/testing";
import { HttpClientTestingModule } from "@angular/common/http/testing";
import { MockStore, provideMockStore } from "@ngrx/store/testing";
import { Film } from "../../models/film.model";
import { FilmsState, getFilmDetail, getFilmsFeatureState, loadFilm, unloadFilm } from "../../store";
import { PosterComponent } from "../../components/poster/poster.component";
import { CardDetailComponent } from "../../components/card-detail/card-detail.component";
import { TranslateModule } from "@ngx-translate/core";
import { getQueryParams, getRouteParams } from "../../../../core/store";
import { Actions } from "@ngrx/effects";
import { Observable } from "rxjs";
import { Action } from "@ngrx/store";
import { provideMockActions } from "@ngrx/effects/testing";
import { MatCardModule } from "@angular/material/card";
import { film } from "../../models/film.fixture";

describe('FilmDetailComponent', () => {
  let component: FilmDetailComponent;
  let fixture: ComponentFixture<FilmDetailComponent>;

  let store: MockStore<FilmsState>;
  let actions$ = new Observable<Action>();

  beforeEach(() => {
    jest.clearAllMocks();

    TestBed.configureTestingModule({
      declarations: [ FilmDetailComponent, PosterComponent, CardDetailComponent ],
      imports: [
        RouterTestingModule,
        HttpClientTestingModule,
        TranslateModule.forRoot({}),
        MatCardModule,
        ],
      providers: [
        provideMockStore(),
        provideMockActions(() => actions$),
        ]
    })
    store = TestBed.inject(MockStore)
    jest.spyOn(store, 'dispatch')

    fixture = TestBed.createComponent(FilmDetailComponent);
    component = fixture.componentInstance;
    actions$ = TestBed.inject(Actions);

  });
  afterEach(() => {
    document.getElementsByTagName('html')[0].innerHTML = '';
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot()
  });
  it('should show details from one movie', () => {
    getFilmDetail.setResult(film);
    store.refreshState()
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot()
  });
  it('should be call LoadFilm with id router and unloadFilm when component destroy', () => {
    getRouteParams.setResult({id: 'test'})
    getQueryParams.setResult({})
    getFilmsFeatureState.setResult({
      details: {
        loading: false,
        film,
        uuid: null
      }, films: {
        films: [],
        loading: false,
      }
    })
    store.refreshState()

    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(loadFilm({uuid: 'test'}))

    fixture.destroy()
    expect(store.dispatch).toHaveBeenCalledWith(unloadFilm())
  })
});
