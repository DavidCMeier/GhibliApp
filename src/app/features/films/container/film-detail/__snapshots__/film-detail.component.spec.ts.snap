// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmDetailComponent should create 1`] = `
<app-film-detail
  __ngContext__={[Function LRootView]}
  film$={[Function Store]}
  loading$={[Function Store]}
  store$={[Function MockStore]}
>
</app-film-detail>
`;

exports[`FilmDetailComponent should show details from one movie 1`] = `
<app-film-detail
  __ngContext__={[Function LRootView]}
  film$={[Function Store]}
  loading$={[Function Store]}
  store$={[Function MockStore]}
>
  <div
    class="flex-content"
  >
    <app-card-detail>
      <div
        class="flex-content"
      >
        <mat-card
          class="mat-card mat-focus-indicator card-detail"
        >
          <mat-card-title
            class="mat-card-title"
          >
            test
          </mat-card-title>
          <mat-card-subtitle
            class="mat-card-subtitle"
          >
             test - test
          </mat-card-subtitle>
          <img
            alt="test banner"
            class="mat-card-image"
            mat-card-image=""
            src="test"
          />
          <mat-card-content
            class="mat-card-content"
          >
            <p>
              films.detail.original_title
            </p>
            <p>
              films.detail.original_title_romanised
            </p>
            <p>
              test
            </p>
            s 
            <p>
              films.detail.directed
            </p>
            <p>
              films.detail.produced
            </p>
            <p>
              films.detail.duration
            </p>
            <p>
              films.detail.rt_score
            </p>
          </mat-card-content>
        </mat-card>
      </div>
    </app-card-detail>
    <app-poster>
      <img
        alt="films.poster.alt"
        class="poster"
        src="test"
      />
    </app-poster>
  </div>
</app-film-detail>
`;
